### DISCUSSION TITLE
How can i run play framework with metals + nvim?

### DISCUSSION THREAD

**Discussion by honeydev** [scalameta/nvim-metals#503] [Q&A]
I find solution for [vscode](https://scalameta.org/metals/blog/2020/08/19/lithium/#running-main-classes-from-dependencies) and try do some for DAP config.

```
{
    type = "scala",
    request = "launch",
    name = "play",
    metals = {
      mainClass = "play.core.server.ProdServerStart",
      buildTarget = "root",
      runType = "run",
      args = {},
      jvmOptions = {}
    }
  }

```
but it's not working for me, I get an error when try run config
`LSP[metals] No main could be found in build target: root`

**Comment by ckipp01** [comment]
Hey thanks for the question @honeydev. You'll want to double check just to make sure in your root project you do indeed have a main method located at `play.core.server.ProdServerStart`. If so, then I'm assuming you'll also see a `run | debug` code lens there? Also do you reproducible example with this? It's quite hard to help without seeing an example. Also, what are you doing with the table you're showing me, are you including it into `require("dap").configurations.scala`?

**Reply by honeydev** [reply]
Thanks for your answer. I goto ProdServerStart definition, no run ui elements.
<img width="1454" alt="image" src="https://user-images.githubusercontent.com/20888582/205518521-26e6ba3c-dbfe-4119-98bf-b0eac4803b7f.png">

In archive app (just default new play project generation)  + full vim config
[play-test.zip](https://github.com/scalameta/nvim-metals/files/10149587/play-test.zip)
MetalsShowBuildTargetInfo ouput
```
Choose build target to view:
1: play-test-build
2: play-test-build-build
3: root
4: root-test
```

Also tests run and work fine.

**Reply by ckipp01** [reply]
Ah, so looking closer at this, this isn't expected to work like this. This main method is in your dependency, not your application. So Metals isn't able to "run" it in this sense. You'll want to use sbt for example to run the actual server.

Regarding tests, I was able to correctly run them from your `HomeControllerSpec`, but seems to be failing loading up some cache. I'm not familiar enough with Play! to understand what should be happening here. However also just running `sbt test` fails for me as well with the same error.
<img width="1386" alt="Screenshot 2022-12-06 at 12 10 01" src="https://user-images.githubusercontent.com/13974112/205896349-a71c2c38-eae2-4444-af81-458ebffa47b9.png">

**Reply by honeydev** [reply]
Yes, this play feature -- run not from main method into you app create this problem. ☹️

**Comment by b2l** [comment]
For anyone landing here with a Java-only (or Java-heavy) Play Framework project: Metals
   won't cut it. Java support is minimal — no completions, no hover, no organize imports
  (https://github.com/scalameta/metals/issues/1815).

  What works: JDTLS + sbt-eclipse. sbt-eclipse generates .classpath/.project files that
  JDTLS consumes natively. Full completions, go-to-definition across modules, diagnostics
   — everything.

  I wrote up the full setup (multi-module, codegen, root detection pitfalls):
 https://dev.to/b2l/neovim-java-lsp-on-a-play-framework-sbt-project-the-missing-guide-1749
